NET-specifics/unit-and-integration-testing.md

```diff
@@ -0,0 +1,171 @@
+
+Apart from this `Context()` that would then be called (and if necessary also overriden) for every test case which inherits from this abstract class, we can also create a GlobalContext() override, where we instantiate resource-heavy objects that are used overall in the test class, since objects in the GlobalContext() are only instantiated once per test class.
+
+
+## The individual test cases
+
+Continuing with our example, we write an individual unit test class for `NewClass`:
+
+```
+public class When_changing_a_NewClass_property_value : concern_for_NewClass
+{
+    protected override void Context()
+    {
+        base.Context();
+        .
+        .
+        .
+    }
+
+    protected override void Because()
+    {
+        ...
+    }
+
+    [Observation]
+    public void should_have_updated_the_value_correctly()
+    {
+        ...
+    }
+
+    [Observation]
+    public void should_notify_about_the_change()
+    {
+        ...
+    }
+}
+```
+
+Please note the convention of naming the unit test classes using complete sentences in lowercase separated with underscores instead of whitespaces. The BDDHelper then replaces the underscores with whitespaces for the test reports. The same naming logic applies to the Observations. Unit test classes usually ( but of course not necessarily ) start with "When_.." and Observations with "should_...". It is important that both the unit test class name as well as the observation fully and correctly describe the behaviour that gets tested and the expected outcome. The length of the name is in this case of no big concern.
+
+### Because() and Observations
+
+Instead of packing both the behaviour that leads to a result that we want to test and the assertion of the correct outcome in one code segment, we prefer to separate those two in a `Because()` function where we have the behaviour that will be tested and an `Observation` that tests the outcome.
+
+
+# Creation of Objects and Mocking 
+
+It is suggested to avoid mocking as far as possible. In our experience this can lead to a situation where too much mocking results in a green test where the actual functionality is not correct and also to more difficulty in maintenance. Instead, it is generally suggested to create real objects where possible and ideally to use helper functions for this task, in order to make the creation of the real objects reusable and avoid code duplication.
+
+Before writing your own function for the creation of an object necessary for testing, please note that in the [HelpersForSpecs](https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/tree/develop/tests/OSPSuite.HelpersForTests) folder you can find functions for the creations of many frequently used objects, like e.g. individuals or simulations. When writing a function to create an object that could be needed overall in the tests, it is these classes exactly that should be extended.
```

Comment:
Here I would also mention application specific Test Helpers - particularly **DomainHelperForSpecs** in PK-Sim and MoBi and other test helper functions:
https://github.com/Open-Systems-Pharmacology/MoBi/tree/develop/tests/MoBi.Tests/Helpers
![grafik](https://github.com/Open-Systems-Pharmacology/developer-docs/assets/25061876/94e8a7ac-4fdf-4347-be9d-8f58d6251902)

Reply:
added